h2o-algos/src/main/java/hex/rulefit/Rule.java

```diff
@@ -0,0 +1,170 @@
+            } else {
+                type = Condition.Type.Categorical;
+                operator = Condition.Operator.In;
+                BitSet inclusiveLevels = node.getRightChild().getInclusiveLevels();
+                List<Integer> matchedDomainValues = new ArrayList<>();
+                String[] catThreshold = new String[inclusiveLevels.cardinality()];
+                int[] catThresholdNum = new int[inclusiveLevels.cardinality()];
+                for (int i = inclusiveLevels.nextSetBit(0); i >= 0; i = inclusiveLevels.nextSetBit(i+1)) {
+                    matchedDomainValues.add(i);
+                }
+                for (int i = 0; i < catThreshold.length; i++) {
+                    catThreshold[i] = node.getDomainValues()[matchedDomainValues.get(i)];
+                    catThresholdNum[i] = matchedDomainValues.get(i);
+                }
+                traverseNodes(node.getRightChild(), new ArrayList<>(conditions), rules, 
+                        new Condition(node.getColId(), type, operator, -1, catThreshold, catThresholdNum, node.getColName(), node.getRightChild().isInclusiveNa()), modelId);
+            }
+               
+            // extract condition leading to the left and recurse there
+            if (node.getDomainValues() == null) {
+                type = Condition.Type.Numerical;
+                operator = Condition.Operator.LessThan;
+                traverseNodes(node.getLeftChild(), new ArrayList<>(conditions), rules, 
+                        new Condition(node.getColId(), type, operator, node.getSplitValue(), null, null, node.getColName(), node.getLeftChild().isInclusiveNa()), modelId);
+            } else {
+                type = Condition.Type.Categorical;
+                operator = Condition.Operator.In;
+                BitSet inclusiveLevels = node.getLeftChild().getInclusiveLevels();
+                List<Integer> matchedDomainValues = new ArrayList<>();
+                String[] catThreshold = new String[inclusiveLevels.cardinality()];
+                int[] catThresholdNum = new int[inclusiveLevels.cardinality()];
+                for (int i = inclusiveLevels.nextSetBit(0); i >= 0; i = inclusiveLevels.nextSetBit(i+1)) {
+                    matchedDomainValues.add(i);
+                }
+                for (int i = 0; i < catThreshold.length; i++) {
+                    catThreshold[i] = node.getDomainValues()[matchedDomainValues.get(i)];
+                    catThresholdNum[i] = matchedDomainValues.get(i);
+                }
+                traverseNodes(node.getLeftChild(), new ArrayList<>(conditions), rules, 
+                        new Condition(node.getColId(), type, operator, -1, catThreshold, catThresholdNum, node.getColName(), node.getLeftChild().isInclusiveNa()), modelId);
+            }
+        }
+    }
+
+    double getAbsCoefficient() {
+        return Math.abs(coefficient);
+    }
+
+    class RuleReducer extends MRTask<RuleReducer> {
```

Comment:
class can be static, less serialisation to do

Reply:
fixed 63d7852